File: 백준/Gold/2573. 빙산/빙산.java
import java.io.*;
import java.util.*;

// 한 덩어리 빙산이 주어질 때, 두 덩어리 이상으로 분리되는 최초 시간?
/**
 * 최대 10년걸림
 * 0. 매 년마다 루프
 * 1. 둘러싼 빙하에 맞게 높이를 update
 *  - 근접 빙하가 영향을 주면 안 됨. 아래 빙하가 3 -> 0 바뀌고, 인접한 내가 3을 0으로 취급하면 안 됨
 * 2. 빙하의 분리 유무 파악
 *  - 분리 되었을 시, 상태 update 및 loop break;
 * 3. 출력
 */

/**
 * 실수한 부분
 * 모두 녹았을 때 == 0이든 <0이든 일관성 있게 저장해야 했는데 그렇지 못함.
 * 문제 조건을 잘못 읽어서 최대 10년이라고 오해함
 */
public class Main {

    static int n;
    static int m;
    static int[][] board = new int[302][302];
    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};

    public static void update() {
        Deque<Change> stack = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 0) continue;

                int cnt = 0;
                for (int k = 0; k < 4; k++) {
                    int nx = i + dx[k];
                    int ny = j + dy[k];

                    if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                    if (board[nx][ny] <= 0) cnt++;
                }

                if (cnt != 0) stack.push(new Change(i, j, Math.max(board[i][j] - cnt, 0)));
            }
        }

        while (!stack.isEmpty()) {
            var cur = stack.pop();
            board[cur.x][cur.y] = cur.val;
        }
    }

    public static boolean isSeparated() {
        int cnt = 0;
        boolean[][] visited = new boolean[302][302];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 0 || visited[i][j]) continue;
                Queue<Pair> queue = new ArrayDeque<>();
                queue.offer(new Pair(i, j));
                visited[i][j] = true;

                while (!queue.isEmpty()) {
                    var cur = queue.poll();

                    for (int k = 0; k < 4; k++) {
                        int nx = cur.x + dx[k];
                        int ny = cur.y + dy[k];

                        if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                        if (board[nx][ny] == 0 || visited[nx][ny]) continue;
                        visited[nx][ny] = true;
                        queue.offer(new Pair(nx, ny));
                    }
                }
                cnt++;
                if (cnt >= 2) return true;
            }
        }
        return false;
    }

    public static boolean allMelted() {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] > 0) return false;
            }
        }
        return true;
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int takeTime = 0;
        while (true) {
            takeTime++;
            update();
            if (isSeparated()) {
                break;
            }

            if (allMelted()) {
                takeTime = 0;
                break;
            }
        }
        System.out.print(takeTime);
    }

    static class Change {
        int x;
        int y;
        int val;

        public Change(int xx, int yy, int vv) {
            x = xx;
            y = yy;
            val = vv;
        }
    }

    static class Pair {
        int x;
        int y;
        public Pair(int xx, int yy) {
            x = xx;
            y = yy;
        }
    }
}

File: 백준/Gold/2573. 빙산/빙산.kt
import kotlin.math.*
import java.io.*
import java.util.*

/**
 * 문제: bfs를 돌리면서 영역의 개수가 2개 이상이 되는 시점 구하기
 * 조건
 * - 각 노드의 값이 1 사이클마다 감소할 수 있음
 * - 최대 영역의 크기: 300
 * 
 * 한 사이클에서의 동작
 * - bfs를 돌려서 영역의 수 구하기: 90,000
 * - 영역이 2개 이상인지 / 0개 인지 검사 -> 탈출
 * - 4개의 주변 블록을 살펴서 다음 사이클의 상태 저장(90,000)
 *  - 다른 배열에 mirroring
 * - 사이클++
 */

val dx = listOf(1, 0, -1, 0)
val dy = listOf(0, 1, 0, -1)

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    var board1 = Array(n) { 
        StringTokenizer(readLine()).run() {
            IntArray(m) { nextToken().toInt() }
        }
    }

    var board2 = Array(n) { IntArray(m) }
    var useFirst = true
    var cycle = 0
    while (true) {
        var visited = Array(n) { BooleanArray(m) }
        val board = if (useFirst) board1 else board2
        val mirror = if (useFirst) {
            board2 = Array(n) { IntArray(m) }
            board2
        } else {
            board1 = Array(n) { IntArray(m) }
            board1
        }
        
        var cnt = 0
        for (i in 0..n - 1) {
            for (j in 0..m - 1) {
                if (board[i][j] == 0 || visited[i][j]) continue
                val queue: Queue<Pair<Int, Int>> = ArrayDeque()
                queue.offer(Pair(i, j))
                visited[i][j] = true

                while (queue.isNotEmpty()) {
                    val (cx, cy) = queue.poll()
                    var adjCount = 0

                    for (k in 0..3) {
                        val nx = cx + dx[k]
                        val ny = cy + dy[k]

                        if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue
                        if (visited[nx][ny]) continue
                        if (board[nx][ny] == 0) {
                            adjCount++
                            continue
                        }
                        visited[nx][ny] = true
                        queue.offer(Pair(nx, ny))
                    }

                    mirror[cx][cy] = 
                        if (board[cx][cy] - adjCount >= 0) board[cx][cy] - adjCount else 0
                }
                cnt++
            }
        }

        if (cnt == 0) {
            cycle = 0
            break
        }

        if (cnt >= 2) {
            break
        }

        cycle++
        useFirst = !useFirst
    }
    print(cycle)
}